src/main/java/org/spongepowered/api/item/recipe/crafting/ShapedCraftingRecipe.java

```diff
@@ -0,0 +1,229 @@
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.item.recipe.crafting;
+
+import org.spongepowered.api.Sponge;
+import org.spongepowered.api.item.ItemTypes;
+import org.spongepowered.api.item.inventory.ItemStack;
+import org.spongepowered.api.item.inventory.ItemStackSnapshot;
+import org.spongepowered.api.item.inventory.Slot;
+import org.spongepowered.api.item.inventory.type.GridInventory;
+import org.spongepowered.api.util.ResettableBuilder;
+import org.spongepowered.api.world.World;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Predicate;
+
+import javax.annotation.Nullable;
+
+/**
+ * A ShapedCraftingRecipe is a CraftingRecipe that has shape and fits into a grid.
+ */
+public interface ShapedCraftingRecipe extends CraftingRecipe {
+    static Builder builder() {
+        return Sponge.getRegistry().createBuilder(Builder.class);
+    }
+
+    /**
+     * @return An unmodifiable copy of the original aisle
+     */
+    List<String> getAisle();
+
+    /**
+     * @return All ingredient predicates; if all of these are fulfilled, the recipe can be crafted
+     */
+    Map<Character, Predicate<ItemStackSnapshot>> getIngredientPredicates();
+
+    /**
+     * @param symbol The symbol paired with the ingredient predicate in the aisle
+     * @return The paired ingredient predicate
+     */
+    default Optional<Predicate<ItemStackSnapshot>> getIngredientPredicate(char symbol) {
+        return Optional.ofNullable(getIngredientPredicates().get(symbol));
+    }
+
+    /**
+     * @param x The x coordinate counted from the left side
+     * @param y The y coordinate counted from the top
+     * @return The ingredient predicate at this position defined by the aisle
+     */
+    default Optional<Predicate<ItemStackSnapshot>> getIngredientPredicate(int x, int y) {
```

Comment:
I'd find it incredibly useful to not have to write these myself, what is the issue exactly? All ShapedCraftingRecipes should behave like this, this is why they are default-implemented. And in corner cases, it also makes it possible to override them. It sounds like a win-win scenario to me. Note that all CraftingRecipes have an `isValid` method.